// @nuxt/devtools/dist/shared/devtools.a2798b81.mjs
import { defineNuxtModule } from '@nuxt/kit';
import { isGlobalInstall } from '../dirs.mjs';

const ROUTE_PATH = "/__nuxt_devtools__";
const ROUTE_ENTRY = `${ROUTE_PATH}/entry`;
const ROUTE_CLIENT = `${ROUTE_PATH}/client`;
const ROUTE_ANALYZE = `${ROUTE_PATH}/analyze`;
const defaultOptions = {
  enabled: void 0,
  // determine multiple conditions
  componentInspector: true,
  viteInspect: true,
  vscode: {
    enabled: true,
    startOnBoot: false,
    port: 3080,
    reuseExistingServer: true
  }
};

const module = defineNuxtModule({
  meta: {
    name: "@nuxt/devtools",
    configKey: "devtools"
  },
  defaults: defaultOptions,
  setup(options, nuxt) {
    if (typeof options === "boolean")
      options = { enabled: options };
    if (options.enabled === false)
      return;
    if (isGlobalInstall()) {
      const globalOptions = nuxt.options.devtoolsGlobal || {};
      if (options.enabled !== true && !globalOptions.projects?.includes(nuxt.options.rootDir))
        return;
    }
    return import('../chunks/module-main.mjs').then(function (n) { return n.m; }).then(({ enableModule }) => enableModule(options, nuxt));
  }
});

export { ROUTE_ENTRY as R, ROUTE_ANALYZE as a, ROUTE_CLIENT as b, module as m };
